firedrake/preconditioners/fdm.py

```diff
@@ -465,678 +490,357 @@ def assemble_coef(self, J, quad_deg, discard_mixed=False, cell_average=False, ne
 
             Q = firedrake.TensorFunctionSpace(mesh, ele, shape=G.ufl_shape)
             q = firedrake.TestFunction(Q)
             Gq_facet = firedrake.Function(Q)
             coefficients["Gq_facet"] = Gq_facet
-            assembly_callables.append(partial(firedrake.assemble, ((inner(q('+'), G('+')) + inner(q('-'), G('-')))/area) * dS_int, Gq_facet))
+            assembly_callables.append(partial(firedrake.assemble, ((inner(q('+'), G('+')) + inner(q('-'), G('-')))/area)*dS_int, Gq_facet))
 
             PT = Piola.T
             Q = firedrake.TensorFunctionSpace(mesh, ele, shape=PT.ufl_shape)
             q = firedrake.TestFunction(Q)
             PT_facet = firedrake.Function(Q)
             coefficients["PT_facet"] = PT_facet
-            assembly_callables.append(partial(firedrake.assemble, ((inner(q('+'), PT('+')) + inner(q('-'), PT('-')))/area) * dS_int, PT_facet))
+            assembly_callables.append(partial(firedrake.assemble, ((inner(q('+'), PT('+')) + inner(q('-'), PT('-')))/area)*dS_int, PT_facet))
         return coefficients, assembly_callables
 
-    @staticmethod
-    def numpy_to_petsc(A_numpy, dense_indices, diag=True):
-        # Create a SeqAIJ Mat from a dense matrix using the diagonal and a subset of rows and columns
-        # If dense_indices is empty, then also include the off-diagonal corners of the matrix
-        n = A_numpy.shape[0]
-        nbase = int(diag) + len(dense_indices)
-        nnz = numpy.full((n,), nbase, dtype=PETSc.IntType)
-        if dense_indices:
-            nnz[dense_indices] = n
-        else:
-            nnz[[0, -1]] = 2
-
-        imode = PETSc.InsertMode.INSERT
-        A_petsc = PETSc.Mat().createAIJ(A_numpy.shape, nnz=nnz, comm=PETSc.COMM_SELF)
-        if diag:
-            for j, ajj in enumerate(A_numpy.diagonal()):
-                A_petsc.setValue(j, j, ajj, imode)
-
-        if dense_indices:
-            idx = numpy.arange(n, dtype=PETSc.IntType)
-            for j in dense_indices:
-                A_petsc.setValues(j, idx, A_numpy[j], imode)
-                A_petsc.setValues(idx, j, A_numpy[:][j], imode)
-        else:
-            A_petsc.setValue(0, n-1, A_numpy[0][-1], imode)
-            A_petsc.setValue(n-1, 0, A_numpy[-1][0], imode)
-
-        A_petsc.assemble()
-        return A_petsc
-
-    @staticmethod
-    def semhat(N, Nq):
-        """
-        Setup for the GLL finite element method on the reference interval
 
-        :arg N: polynomial degree of the GLL element
-        :arg Nq: quadrature degree (Nq >= 2*N+1 gives exact integrals)
+def get_piola_tensor(mapping, domain):
+    tdim = domain.topological_dimension()
+    if mapping == 'identity':
+        return None
+    elif mapping == 'covariant piola':
+        return ufl.JacobianInverse(domain).T * firedrake.Constant(numpy.flipud(numpy.identity(tdim)), domain=domain)
```

Comment:
For these actually constant matrices, you could just use `ufl.as_tensor(some_literal_array)` rather than `firedrake.Constant`.